// @carto/api-client/src/sources/index.ts
// deck.gl
// SPDX-License-Identifier: MIT
// Copyright (c) vis.gl contributors

export {SOURCE_DEFAULTS} from './base-source';
export type {
  GeojsonResult,
  JsonResult,
  QueryResult,
  QuerySourceOptions,
  RasterBandColorinterp,
  RasterMetadata,
  RasterMetadataBand,
  RasterMetadataBandStats,
  SourceOptions,
  SpatialFilterPolyfillMode,
  TableSourceOptions,
  TilejsonResult,
  TileResolution,
  TilesetSourceOptions,
  VectorLayer,
} from './types';

export {boundaryQuerySource} from './boundary-query-source';
export type {
  BoundaryQuerySourceOptions,
  BoundaryQuerySourceResponse,
} from './boundary-query-source';

export {boundaryTableSource} from './boundary-table-source';
export type {
  BoundaryTableSourceOptions,
  BoundaryTableSourceResponse,
} from './boundary-table-source';

export {h3QuerySource} from './h3-query-source';
export type {
  H3QuerySourceOptions,
  H3QuerySourceResponse,
} from './h3-query-source';

export {h3TableSource} from './h3-table-source';
export type {
  H3TableSourceOptions,
  H3TableSourceResponse,
} from './h3-table-source';

export {h3TilesetSource} from './h3-tileset-source';
export type {
  H3TilesetSourceOptions,
  H3TilesetSourceResponse,
} from './h3-tileset-source';

export {rasterSource} from './raster-source';
export type {RasterSourceOptions, RasterSourceResponse} from './raster-source';

export {quadbinQuerySource} from './quadbin-query-source';
export type {
  QuadbinQuerySourceOptions,
  QuadbinQuerySourceResponse,
} from './quadbin-query-source';

export {quadbinTableSource} from './quadbin-table-source';
export type {
  QuadbinTableSourceOptions,
  QuadbinTableSourceResponse,
} from './quadbin-table-source';

export {quadbinTilesetSource} from './quadbin-tileset-source';
export type {
  QuadbinTilesetSourceOptions,
  QuadbinTilesetSourceResponse,
} from './quadbin-tileset-source';

export {vectorQuerySource} from './vector-query-source';
export type {
  VectorQuerySourceOptions,
  VectorQuerySourceResponse,
} from './vector-query-source';

export {vectorTableSource} from './vector-table-source';
export type {
  VectorTableSourceOptions,
  VectorTableSourceResponse,
} from './vector-table-source';

export {vectorTilesetSource} from './vector-tileset-source';
export type {
  VectorTilesetSourceOptions,
  VectorTilesetSourceResponse,
} from './vector-tileset-source';
